<!DOCTYPE html>
<html lang="en">

<!-- <head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1">

  <title>Scribe</title>
  <meta name="description" content="Scribe - Gene regulation visualization, causal networkinference for single cell measurements">

  <link rel="stylesheet" href="/scribe/css/main.css">
  <link rel="canonical" href="http://localhost:4000/scribe/features.html">
  <link rel="alternate" type="application/rss+xml" title="Scribe" href="/scribe/feed.xml">

  
</head> -->

<head>
   <meta charset="utf-8">
   <meta http-equiv="X-UA-Compatible" content="IE=edge">
   <meta name="viewport" content="width=device-width, initial-scale=1">
   <!-- The above 3 meta tags *must* come first in the head; any other head content must come *after* these tags -->
   <title>Scribe</title>
   <meta name="description" content="Scribe - Gene regulation visualization, causal networkinference for single cell measurements">
   <!-- <link rel="icon" href="../../favicon.ico"> -->

   <title>Theme Template for Bootstrap</title>

  <link rel="stylesheet" href="/scribe/css/main.css">
  <link rel="stylesheet" href="/scribe/node_modules/prismjs/themes/base16-atelierlakeside.light.css" />
  <link rel="stylesheet" href="https://cdn.rawgit.com/afeld/bootstrap-toc/v0.4.1/dist/bootstrap-toc.min.css">
  <!-- add after bootstrap.min.js -->
  <link rel="canonical" href="http://localhost:4000/scribe/features.html">


   <!-- HTML5 shim and Respond.js for IE8 support of HTML5 elements and media queries -->
   <!--[if lt IE 9]>
     <script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
     <script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
   <![endif]-->

   <!--  -->
 </head>


<body>
  <!-- <header class="site-header" role="banner"> -->
  <div class="container-fullwidth">
  <nav class="navbar navbar-inverse navbar-static-top">
    <div class="container">
      <div class="navbar-header">
        <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false" aria-controls="navbar">
          <span class="sr-only">Toggle navigation</span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
        </button>
        <a class="navbar-brand" href="/scribe/">Scribe</a>
      </div>
      <div id="navbar" class="collapse navbar-collapse">
        <ul class="nav navbar-nav">
          <!-- <li class="active"><a href="#">Home</a></li> -->

          <script>

          function mobileDetect() {
            if (/Android|webOS|iPhone|iPad|iPod|BlackBerry|BB|PlayBook|IEMobile|Windows Phone|Kindle|Silk|Opera Mini/i.test(navigator.userAgent)) {
              window.location.href = "/scribe/docs_mobile/"
            } else {
              window.location.href = "/scribe/docs/"
            }
          }

          </script>

          <li><a href="javascript:void(0)" onclick="mobileDetect();">Documentation</a></li>
          <!-- <li><a href="/features/">Features</a></li> -->
          <!-- <li><a href="/getting-started/">Getting started</a></li> -->

          <li><a href="/scribe/papers/">Publications</a></li>
          <li><a href="/scribe/tutorials/">Tutorials</a></li>
          <li><a href="https://github.com/cole-trapnell-lab/monocle-release">GitHub</a></li>
        </ul>
      </div><!--/.nav-collapse -->
    </div>
  </nav>
  </div>
  <!-- <div class="wrapper">
    
    
    <a class="site-title" href="/scribe/">Scribe</a>

    
      <nav class="site-nav">
        <input type="checkbox" id="nav-trigger" class="nav-trigger" />
        <label for="nav-trigger">
          <span class="menu-icon">
            <svg viewBox="0 0 18 15" width="18px" height="15px">
              <path fill="#424242" d="M18,1.484c0,0.82-0.665,1.484-1.484,1.484H1.484C0.665,2.969,0,2.304,0,1.484l0,0C0,0.665,0.665,0,1.484,0 h15.031C17.335,0,18,0.665,18,1.484L18,1.484z"/>
              <path fill="#424242" d="M18,7.516C18,8.335,17.335,9,16.516,9H1.484C0.665,9,0,8.335,0,7.516l0,0c0-0.82,0.665-1.484,1.484-1.484 h15.031C17.335,6.031,18,6.696,18,7.516L18,7.516z"/>
              <path fill="#424242" d="M18,13.516C18,14.335,17.335,15,16.516,15H1.484C0.665,15,0,14.335,0,13.516l0,0 c0-0.82,0.665-1.484,1.484-1.484h15.031C17.335,12.031,18,12.696,18,13.516L18,13.516z"/>
            </svg>
          </span>
        </label>

        <div class="trigger">
          
            
            
          
            
            
          
            
            
          
            
            
          
            
            
          
            
            
          
            
            
          
            
            
          
            
            
          
            
            
          
            
            
          
            
            
          
            
            
          
            
            
          
            
            
          
            
            
            <a class="page-link" href="/scribe/node_modules/bootstrap-sass/CHANGELOG.html">Changelog</a>
            
          
            
            
            <a class="page-link" href="/scribe/node_modules/prismjs/CHANGELOG.html">Prism Changelog</a>
            
          
            
            
            <a class="page-link" href="/scribe/node_modules/bootstrap-sass/">Bootstrap for Sass</a>
            
          
            
            
            <a class="page-link" href="/scribe/node_modules/clipboard/">clipboard.js</a>
            
          
            
            
            <a class="page-link" href="/scribe/node_modules/delegate/">delegate</a>
            
          
            
            
            <a class="page-link" href="/scribe/node_modules/good-listener/">good-listener</a>
            
          
            
            
            <a class="page-link" href="/scribe/node_modules/jquery/">jQuery</a>
            
          
            
            
            <a class="page-link" href="/scribe/node_modules/prismjs/">Prism</a>
            
          
            
            
            <a class="page-link" href="/scribe/node_modules/select/">select</a>
            
          
            
            
            <a class="page-link" href="/scribe/node_modules/tiny-emitter/">tiny-emitter</a>
            
          
            
            
          
        </div>
      </nav>
    
  </div> -->
</header>


  <div class="container marketing">
  <!-- Carousel
  ================================================== -->
  <!-- <div id="myCarousel" class="carousel slide" data-ride="carousel">

    <ol class="carousel-indicators">
      <li data-target="#myCarousel" data-slide-to="0" class="active"></li>
      <li data-target="#myCarousel" data-slide-to="1"></li>
      <li data-target="#myCarousel" data-slide-to="2"></li>
    </ol>
    <div class="carousel-inner" role="listbox">
      <div class="item active">
        <div class="text-center">
        <img class="first-slide" src="/scribe/images/cell_type_impute.png" alt="First slide">
      </div>
        <div class="container">
          <div class="carousel-caption">
            <h1>Example headline.</h1>
            <p>Note: If you're viewing this page via a <code>file://</code> URL, the "next" and "previous" Glyphicon buttons on the left and right might not load/display properly due to web browser security rules.</p>
            <p><a class="btn btn-lg btn-primary" href="#" role="button">Sign up today</a></p>
          </div>
        </div>
      </div>
      <div class="item">
        <img class="second-slide" src="/scribe/images/all_cells_marker_heatmap.png" alt="Second slide">
        <div class="container">
          <div class="carousel-caption">
            <h1>Another example headline.</h1>
            <p>Cras justo odio, dapibus ac facilisis in, egestas eget quam. Donec id elit non mi porta gravida at eget metus. Nullam id dolor id nibh ultricies vehicula ut id elit.</p>
            <p><a class="btn btn-lg btn-primary" href="#" role="button">Learn more</a></p>
          </div>
        </div>
      </div>
      <div class="item">
        <img class="third-slide" src="/scribe/images/branch_trajectory.png" alt="Third slide">
        <div class="container">
          <div class="carousel-caption">
            <h1>One more for good measure.</h1>
            <p>Cras justo odio, dapibus ac facilisis in, egestas eget quam. Donec id elit non mi porta gravida at eget metus. Nullam id dolor id nibh ultricies vehicula ut id elit.</p>
            <p><a class="btn btn-lg btn-primary" href="#" role="button">Browse gallery</a></p>
          </div>
        </div>
      </div>
    </div>
    <a class="left carousel-control" href="#myCarousel" role="button" data-slide="prev">
      <span class="glyphicon glyphicon-chevron-left" aria-hidden="true"></span>
      <span class="sr-only">Previous</span>
    </a>
    <a class="right carousel-control" href="#myCarousel" role="button" data-slide="next">
      <span class="glyphicon glyphicon-chevron-right" aria-hidden="true"></span>
      <span class="sr-only">Next</span>
    </a>
  </div> -->


  <!-- Marketing messaging and featurettes
  ================================================== -->
  <!-- Wrap the rest of the page in another container to center all the content. -->



    <!-- Three columns of text below the carousel -->
    <div class="row">
      <div class="col-lg-4">
        <img class="img-circle" src="data:image/gif;base64,R0lGODlhAQABAIAAAHd3dwAAACH5BAAAAAAALAAAAAABAAEAAAICRAEAOw==" alt="Generic placeholder image" width="140" height="140">
        <h2>Heading</h2>
        <p>Donec sed odio dui. Etiam porta sem malesuada magna mollis euismod. Nullam id dolor id nibh ultricies vehicula ut id elit. Morbi leo risus, porta ac consectetur ac, vestibulum at eros. Praesent commodo cursus magna.</p>
        <p><a class="btn btn-default" href="#" role="button">View details &raquo;</a></p>
      </div><!-- /.col-lg-4 -->
      <div class="col-lg-4">
        <img class="img-circle" src="data:image/gif;base64,R0lGODlhAQABAIAAAHd3dwAAACH5BAAAAAAALAAAAAABAAEAAAICRAEAOw==" alt="Generic placeholder image" width="140" height="140">
        <h2>Heading</h2>
        <p>Duis mollis, est non commodo luctus, nisi erat porttitor ligula, eget lacinia odio sem nec elit. Cras mattis consectetur purus sit amet fermentum. Fusce dapibus, tellus ac cursus commodo, tortor mauris condimentum nibh.</p>
        <p><a class="btn btn-default" href="#" role="button">View details &raquo;</a></p>
      </div><!-- /.col-lg-4 -->
      <div class="col-lg-4">
        <img class="img-circle" src="data:image/gif;base64,R0lGODlhAQABAIAAAHd3dwAAACH5BAAAAAAALAAAAAABAAEAAAICRAEAOw==" alt="Generic placeholder image" width="140" height="140">
        <h2>Heading</h2>
        <p>Donec sed odio dui. Cras justo odio, dapibus ac facilisis in, egestas eget quam. Vestibulum id ligula porta felis euismod semper. Fusce dapibus, tellus ac cursus commodo, tortor mauris condimentum nibh, ut fermentum massa justo sit amet risus.</p>
        <p><a class="btn btn-default" href="#" role="button">View details &raquo;</a></p>
      </div><!-- /.col-lg-4 -->
    </div><!-- /.row -->


    <!-- START THE FEATURETTES -->

    <hr class="featurette-divider">

    <div class="row featurette">
      <div class="col-md-7 col-md-push-5">
        <h2 class="featurette-heading">Classify and count cells. <span class="text-muted">See for yourself.</span></h2>
        <p class="lead">Single-cell RNA-Seq experiments allow you to discover (and possibly rare) subtypes of cells. Monocle helps you identify them.</p>

        <p>Most single-cell RNA-Seq protocols capture only a small percentage of the mRNA in each cell, which leads to so-called "drop outs", where mRNAs that are probably present just aren't detected. This makes it hard to determine how many cells of each type there are in your experiment. Monocle has a machine-learning based algorithm for classifying cells by type even in the face of missing data, and even when your cells have been treated in different ways or collected from varying conditions. You can use it to track population dynamics, measure responses to perturbation, and even discover new cell types.</p>

        <p>In the plot above, we can recognize some cells as fibroblasts and others as myoblasts on the basis of a few key marker genes. Even though these markers drop-out in many cells, Monocle 2 can still identify cells that lack them by clustering all the cells together based on global expression profile.</p>
      </div>
      <div class="col-md-5 col-md-pull-7">
        <img class="featurette-image img-responsive center-block" src="/scribe/images/cell_type_impute.png" width=300 height=300 alt="Generic placeholder image">
      </div>
    </div>


        <hr class="featurette-divider">

        <div class="row featurette">
          <div class="col-md-7">
            <h2 class="featurette-heading">Compare cell populations through differential expression. <span class="text-muted">It'll blow your mind.</span></h2>
            <p class="lead">Single-cell RNA-Seq promises to unveil new cell types and new functional states for known ones. Characterizing new cell types and states begins with comparing them to other, better understood cells. Monocle includes a sophisticated but easy-to-use system for differential expression.</p>
            <p>Monocle allows you to compare groups of cells to identify genes that differ between them, even after controlling for other variables in the experiment. For example, the plot above shows genes that differ between two conditions, but only one of them, LAG3, is **also** different between cell types, even after subtracting the other effects. You can use Monocle's differential expression system to zero in on the genes that are most relevant in your experiment.</p>
          </div>
          <div class="col-md-5">
            <img class="featurette-image img-responsive center-block" src="/scribe/images/example_T_cell_differential_genes.png" width=300 height=300 alt="Generic placeholder image">
          </div>
        </div>


    <hr class="featurette-divider">

    <div class="row featurette">
      <div class="col-md-7 col-md-push-5">
        <h2 class="featurette-heading">Identify new marker genes. <span class="text-muted">Checkmate.</span></h2>
        <p class="lead">Many researchers are using single-cell RNA-Seq to discover new cell types. Monocle can help you purify them or characterize them further by identifying key marker genes that you can use in follow up experiments such as immunofluorescence or flow sorting.
        <p>Monocle leverages its differential analysis procedures so you can be assured of the statistical robustness of your marker panel, leading to reliable downstream analysis.</p>
        <p> In the plot above, Monocle has identified the top 20 most cell-type specific genes for some of the major lymphocytes in peripheral blood.</p>

      </div>
      <div class="col-md-5 col-md-pull-7">
        <img class="featurette-image img-responsive center-block" src="/scribe/images/all_cells_marker_heatmap.png" height=300 width=300 alt="Generic placeholder image">
      </div>
    </div>


    <hr class="featurette-divider">

    <div class="row featurette">
      <div class="col-md-7">
        <h2 class="featurette-heading">Robustly track changes over (pseudo) time. <span class="text-muted">It'll blow your mind.</span></h2>
        <p class="lead">
        In development, disease, and throughout life, cells transition from one state to another. Monocle helps you discover these transitions.</p>
        <p>
        Monocle introduced the concept of **pseudotime**, which is a measure of how far a cell has moved through biological progress. Monocle uses machine learning to reconstruct the sequence of gene regulatory changes each cell must execute as it transitions from one state to another. It can do so without knowing ahead of time which genes define these states or movement between them. After it learns the transition path, or *trajectory*, it places each cell at the correct position along it.</p>
<p>The plot above shows the transcriptional trajectory from developing [olfactory neurons](http://science.sciencemag.org/content/350/6265/1251). (Hanchate et al, Science, 2015).
<p>Essentially, Monocle treats each cell as a separate time point in an ultra-high resolution time series. Instead of three or four time points, you get three or four thousand - one for each cell in the experiment. This resolution allows you distinguish genes regulated early in a complex cascade of transcriptional changes from those more downstream. For example, in the plot above you can see that Neurod1 is downregulated before cells enter an immature neuronal state, after which Gng8 is upregulated. As the cells reach maturity, Cnga2 becomes highly expressed.

<p>Monocle 2 includes a vastly improved algorithm for learning trajectories and ordering cells. It's based on a recently introduced technique called [reverse graph embedding](http://dl.acm.org/citation.cfm?id=2783309) (Mao et al, KDD 15), a highly scalable nonlinear manifold learning technique.

      </div>
      <div class="col-md-5">
        <img class="featurette-image img-responsive center-block" src="/scribe/images/neuron_mst.png" width=300 height=300 alt="Generic placeholder image">
      </div>
      <div class="col-md-5">
        <img class="featurette-image img-responsive center-block" src="/scribe/images/neuron_pseudotime_changes.png" width=300 height=300 alt="Generic placeholder image">
      </div>
    </div>



    <hr class="featurette-divider">

    <div class="row featurette">
      <div class="col-md-7 col-md-push-5">
        <h2 class="featurette-heading">Dissect cellular decisions with branch analysis. <span class="text-muted">Checkmate.</span></h2>
        <p class="lead">Single-cell trajectory analysis not only exposes transient events as cells move from one state to another. It reveals how cells choose between one of several possible end states. For example, in development, multipotent cells make fate choices, ultimately differentiating into one of a number of terminal cell types. The new reconstruction algorithms introduced in Monocle 2 can robustly reveal branching trajectories. Inspecting the branch points helps you dissect the gene regulatory circuits that cells use to navigate these decisions.
        <p>Here, you can see cells that start at a common state (shown in pink) and move through a branched trajectory to one of several functionally distinct outcomes. Which genes are affected at the branches? Monocle 2 includes a new statistical procedure, called Branched Expression Analysis Modeling (BEAM), that reports all genes that change around a selected branch. BEAM can be used to track genes along two lineages in parallel from both a **global** perspective:

      </div>
      <div class="col-md-5 col-md-pull-7">
        <img class="featurette-image img-responsive center-block" src="/scribe/images/branch_trajectory.png" height=300 width=300 alt="Generic placeholder image">
      </div>

      <div class="col-md-5 col-md-pull-7">
        <img class="featurette-image img-responsive center-block" src="/scribe/images/branched_pseudotime.png" height=300 width=300 alt="Generic placeholder image">
      </div>

    </div>

    <hr class="featurette-divider">

    <!-- /END THE FEATURETTES -->

    <script src="/scribe/node_modules/jquery/dist/jquery.min.js"></script>
<script src="/scribe/node_modules/bootstrap-sass/assets/javascripts/bootstrap.min.js"></script>
<script src="/scribe/node_modules/prismjs/prism.js"></script>
<script src="/scribe/node_modules/prismjs/prism.r.js" type="text/javascript"></script>
<script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v0.4.1/dist/bootstrap-toc.min.js"></script>

<script type="text/x-mathjax-config">
  MathJax.Hub.Config({tex2jax: {inlineMath: [['$','$'], ['\\(','\\)']]}});
</script>
<script type="text/javascript" async
  src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-MML-AM_CHTML">
</script>

<footer class="site-footer">

  <div class="wrapper">

    <!-- <h2 class="footer-heading">Scribe</h2>

    <div class="footer-col-wrapper">
      <div class="footer-col footer-col-1">
        <ul class="contact-list">
          <li>
            
              Scribe
            
            </li>
            
            <li><a href="mailto:coletrap@uw.edu">coletrap@uw.edu</a></li>
            
        </ul>
      </div>

      <div class="footer-col footer-col-2">
        <ul class="social-media-list">
          

          
        </ul>
      </div>

      <div class="footer-col footer-col-3">
        <p>Scribe -  Gene regulation visualization, causal networkinference for single cell measurements
</p>
      </div>
    </div> -->

  </div>

</footer>


  </body>

  </html>

<!--
Monocle is a toolkit for analyzing large, complex single-cell RNA-Seq experiments. It includes functions that can help you tackle the challenges that are common in most such experiments. Recently, we overhauled Monocle to support the next generation of single-cell RNA-Seq. The new improved algorithms in Monocle 2 can handle much larger and complex experiments. With Monocle 2, you can:

<p style="text-align:center"><img src="/scribe/images/neuron_pseudotime_changes.png"/></p>

Essentially, Monocle treats each cell as a separate time point in an ultra-high resolution time series. Instead of three or four time points, you get three or four thousand - one for each cell in the experiment. This resolution allows you distinguish genes regulated early in a complex cascade of transcriptional changes from those more downstream. For example, in the plot above you can see that Neurod1 is downregulated before cells enter an immature neuronal state, after which Gng8 is upregulated. As the cells reach maturity, Cnga2 becomes highly expressed.

# Dissect cellular decisions with branch analysis

Single-cell trajectory analysis not only exposes transient events as cells move from one state to another. It reveals how cells choose between one of several possible end states. For example, in development, multipotent cells make fate choices, ultimately differentiating into one of a number of terminal cell types. The new reconstruction algorithms introduced in Monocle 2 can robustly reveal branching trajectories. Inspecting the branch points helps you dissect the gene regulatory circuits that cells use to navigate these decisions.

<p style="text-align:center"><img src="/scribe/images/branch_trajectory.png"/></p>


Here, you can see cells that start at a common state (shown in pink) and move through a branched trajectory to one of several functionally distinct outcomes. Which genes are affected at the branches? Monocle 2 includes a new statistical procedure, called Branched Expression Analysis Modeling (BEAM), that reports all genes that change around a selected branch. BEAM can be used to track genes along two lineages in parallel from both a **global** perspective:

<p style="text-align:center"><img src="/scribe/images/BEAM_heatmap.png"/></p>

As well as a on a **gene-by-gene** basis:

<p style="text-align:center"><img src="/scribe/images/branched_pseudotime.png"/></p>

In both plots, you can see the progressive lineage restriction of the genes along the two branches as a function of pseudotime. Further analyzing these genes can lead to insights about how the lineage choice is made in development.

# Count transcripts, not reads

In RNA-Seq, we measure expression by counting how many sequencing reads originate from each gene.  But what really matters is how many mRNAs there were for each gene. Spiking in "alien" transcripts such as the [ERCC spike-in controls](https://www.thermofisher.com/order/catalog/product/4456740) allows you to transform expression measurements from read counts into mRNA counts. However, spike-in controls can be tricky to use correctly, and they aren't always reliable. Monocle allows you to analyze mRNA counts *without the need for spike-in controls*. It includes an algorithm motivated by some surprising and fundamental statistical properties of the single-cell mRNA sampling process that lets you recover mRNA counts via inference rather than an exogenous ladder. These counts are generally as or more reliable than those obtained with spikes.

<p style="text-align:center"><img src="/scribe/images/spike_free_consistency.png"/></p>

Here, you can see that mRNA counts from Monocle's spike-free algorithm are actually more consistent across cells over a range of sequencing depths than those from spike-in controls or a measure of relative expression.

# Analyze huge experiments

With current single-cell RNA-Seq techniques such as [Drop-Seq](http://mccarrolllab.com/dropseq/) / [inDrop](http://www.cell.com/cell/abstract/S0092-8674(15)00500-0), you can easily capture thousands or even tens of thousands of cells in a single experiment. Such experiments often include many different conditions and controls meant to probe the structure of a complex cellular community. Analyzing such datasets is extremely challenging, in part because of their sheer size. For example, the plot below is from an experiment with more than 50,000 cells!

<p style="text-align:center"><img src="/scribe/images/cell_counts_10X.png"/></p>

Monocle 2 has been re-engineered from the ground up to support large, sparse RNA-Seq datasets. All of the major steps in the Monocle workflow including cell clustering, differential expression, and trajectory analysis can be performed on a single PC, even for huge datasets.  -->
